### PR TITLE
Add viewing and creating uploads

### PR DESCRIPTION
**Pull request by matthewdias**
🌆  ⬆️

### CODE REVIEW

**Comment by NuckChorris** [comment]
Just rebased this, @vevix @matthewdias do either of you remember what still needed done?

**Comment by matthewdias** [comment]
iirc it was finished unless there are some new bugs or requirements

File: app/components/stream-feed/create-post.js [hunk 1]
```diff
@@ -18,20 +24,25 @@ export default Component.extend({
   spoiler: false,
   shouldUnit: false,
   maxLength: 9000,
+  uploads: [],
```
Comment (trmcnvn) [inline app/components/stream-feed/create-post.js]: Move this initialization to `init`

Comment (trmcnvn) [reply app/components/stream-feed/create-post.js]: Ref: https://poteto.github.io/component-best-practices/#/leaky

File: app/components/stream-feed/create-post.js [hunk 2]
```diff
@@ -82,6 +93,29 @@ export default Component.extend({
...
+      const store = get(this, 'store');
+      store.pushPayload(body);
+      const uploads = get(this, 'uploads');
+      uploads.addObjects(body.data.map(upload => store.peekRecord('upload', upload.id)));
+      this._orderUploads(uploads);
+    } catch (error) {
+      get(this, 'notify').error(errorMessages(error));
+      const queue = get(this, 'fileQueue').find('uploads');
+      get(queue, 'files').forEach(file => set(file, 'queue', null));
+      set(queue, 'files', A());
```
Comment (trmcnvn) [inline app/components/stream-feed/create-post.js]: Capture this exception with Raven

File: app/components/stream-feed/create-post.js [hunk 3]
```diff
@@ -164,6 +204,33 @@ export default Component.extend({
...
+      }
+    },
+
+    reorderUploads(orderedUploads) {
+      this._orderUploads(orderedUploads);
+    },
+
+    removeUpload(upload) {
+      get(this, 'uploads').removeObject(upload);
+    },
```
Comment (trmcnvn) [inline app/components/stream-feed/create-post.js]: Strip `,`

File: app/components/stream-feed/items/post.js [hunk 4]
```diff
@@ -20,6 +21,7 @@ export default Component.extend(ClipboardMixin, CanMixin, {
   isFollowingPost: false,
   isOverflowed: false,
   isExpanded: false,
+  uploads: [],
```
Comment (trmcnvn) [inline app/components/stream-feed/items/post.js]: Same as above comment regarding Object/Array init on components.

File: app/components/stream-feed/items/post.js [hunk 5]
```diff
@@ -64,6 +66,16 @@ export default Component.extend(ClipboardMixin, CanMixin, {
     return this.can('edit post', get(this, 'post'));
   }).readOnly(),
 
+  getUploadsTask: task(function* () {
+    return yield get(this, 'store').query('upload', {
+      filter: {
+        ownerId: get(this, 'post.id'),
+        ownerType: 'Post'
```
Comment (trmcnvn) [inline app/components/stream-feed/items/post.js]: We are sure this doesn't expect lowercase?

File: app/components/stream-feed/items/post/comment-box.js [hunk 6]
```diff
@@ -0,0 +1,81 @@
+import Component from 'ember-component';
+import { isEmpty } from 'ember-utils';
+import get from 'ember-metal/get';
+import set from 'ember-metal/set';
+import service from 'ember-service/inject';
+import { notEmpty } from 'ember-computed';
+import { A } from 'ember-array/utils';
```
Comment (trmcnvn) [inline app/components/stream-feed/items/post/comment-box.js]: Update these to use new module imports.

File: app/components/stream-feed/items/post/comment-box.js [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+        headers
+      });
+      const store = get(this, 'store');
+      store.pushPayload(body);
+      set(this, 'upload', store.peekRecord('upload', body.data[0].id));
+    } catch (error) {
+      get(this, 'notify').error(errorMessages(error));
+      const queue = get(this, 'fileQueue').find(`comment-uploads-${get(this, 'elementId')}`);
+      get(queue, 'files').forEach(file => set(file, 'queue', null));
+      set(queue, 'files', A());
```
Comment (trmcnvn) [inline app/components/stream-feed/items/post/comment-box.js]: Same as above: Capture with Raven and can we use `[]`.

File: app/components/stream-feed/items/post/comment-box.js [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+        reader.addEventListener('load', () => {
+          get(this, 'uploadImageTask').perform(File.fromDataURL(reader.result));
+        });
+        reader.readAsDataURL(image);
+      }
+    },
+
+    removeUpload() {
+      set(this, 'upload', undefined);
+    },
```
Comment (trmcnvn) [inline app/components/stream-feed/items/post/comment-box.js]: ✂️ `,`

File: app/components/stream-feed/items/post/comment.js [hunk 9]
```diff
@@ -14,6 +16,8 @@ import { CanMixin } from 'ember-can';
 
 export default Component.extend(ClipboardMixin, Pagination, CanMixin, {
   classNameBindings: ['comment.isNew:new-comment'],
+  upload: undefined,
+  galleryItems: [],
```
Comment (trmcnvn) [inline app/components/stream-feed/items/post/comment.js]: 👉 `init`

File: app/components/stream-feed/items/post/uploads-grid.js [hunk 10]
```diff
@@ -0,0 +1,138 @@
+import Component from 'ember-component';
+import get, { getProperties } from 'ember-metal/get';
+import set, { setProperties } from 'ember-metal/set';
+import computed from 'ember-computed';
```
Comment (trmcnvn) [inline app/components/stream-feed/items/post/uploads-grid.js]: Update to new module imports.

File: app/components/stream-feed/items/post/uploads-grid.js [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+        short = ((max * 0.83) - margin) * (2 / 5);
+      }
+      break;
+    default:
+  }
+  if (orientation === 'landscape') {
+    return { width: long, height: short };
+  }
+  return { width: short, height: long };
+};
```
Comment (trmcnvn) [inline app/components/stream-feed/items/post/uploads-grid.js:93]: I'm okay with leaving this here, but it's probably best to extract to a util.

File: app/components/stream-feed/items/post/uploads-grid.js [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+
+export default Component.extend({
+  classNames: ['stream-content-post-uploads-container'],
+  classNameBindings: ['gridStyle'],
+  uploads: [],
+  galleryItems: [],
```
Comment (trmcnvn) [inline app/components/stream-feed/items/post/uploads-grid.js]: 👉 `init`

File: app/serializers/comment.js [hunk 13]
```diff
@@ -7,6 +7,8 @@ export default ApplicationSerializer.extend({
     repliesCount: { serialize: false },
     createdAt: { serialize: false },
     editedAt: { serialize: false },
-    updatedAt: { serialize: false }
+    updatedAt: { serialize: false },
+
+    uploads: { serialize: true }
```
Comment (trmcnvn) [inline app/serializers/comment.js]: Is this needed at all? Should serialize all except ones explicitly defined not to.

File: app/serializers/post.js [hunk 14]
```diff
@@ -8,6 +8,8 @@ export default ApplicationSerializer.extend({
     topLevelCommentsCount: { serialize: false },
     createdAt: { serialize: false },
     editedAt: { serialize: false },
-    updatedAt: { serialize: false }
+    updatedAt: { serialize: false },
+
+    uploads: { serialize: true }
```
Comment (trmcnvn) [inline app/serializers/post.js]: ⬆️

File: app/components/stream-feed/create-post.js (hunk 2 above)
Comment (trmcnvn) [inline app/components/stream-feed/create-post.js]: Can we just `[]` here instead of Ember.Array?

**Review by trmcnvn — CHANGES_REQUESTED** [review]

**Review by trmcnvn — APPROVED** [review]

**Pull request merged**